### PR TITLE
TST: Implement a `run_in_parallel` decorator for test functions

### PR DESCRIPTION
**Pull request by andfoy**
#### Reference issue

See https://github.com/Quansight-Labs/free-threaded-compatibility/issues/56

#### What does this implement/fix?

This PR replaces the `_run_concurrent_barrier` introduced in #20671 by an unified decorator that supports pytest fixtures and markers, making it more idiomatic whilst reducing the amount of boilerplate code. This will enable easier concurrency and parallel testing of functions, classes and methods

### CODE REVIEW

**Comment by andfoy** [comment]
cc @rgommers for visibility

File: scipy/_lib/_testutils.py [hunk 1]
```diff
@@ -326,41 +328,104 @@ def load(modname):
...
-        Function closure to execute concurrently. Its first argument will
-        be the thread id.
-    *args: tuple
-        Variable number of positional arguments to pass to the function.
-    **kwargs: dict
-        Keyword arguments to pass to the function.
+        Number of concurrent threads to spawn. Default: 10
+    use_barrier: bool
+        If `True`, then a barrier is set before all threads start executing the
+        target function. Default: `True`
```
Comment (ev-br) [inline scipy/_lib/_testutils.py:393]: What would be the point of using `use_barrier=False` here?

File: scipy/_lib/_testutils.py [hunk 2]
```diff
@@ -326,41 +328,104 @@ def load(modname):
...
+    assert_all_equal: bool
+        If `True`, then the return value produced by the function will be
+        compared for equality across all executing threads. Default: `True`
```
Comment (ev-br) [inline scipy/_lib/_testutils.py:396]: Minor: the name hints that it checks equality but the actual test is `assert_allclose` with the default tolerance. Before addressing this small inconsistency, could you help me understand the intended usages, is it:

- run a bunch of identical workloads concurrently, check that that all results are identical
- run a bunch of possibly different workloads, do not check anything

Is this enough?

**Review by ev-br — COMMENTED** [review]
This PR seems to be an identical refactor of the testing machinery added in https://github.com/scipy/scipy/pull/20671.
IIUC, the machinery was already useful to smoke out several issues with threads.

However, while we're at it, let's maybe start from what testing patterns we actually want to allow. My knee-jerk suggestion would be to check that _something_ produces identical results when run in parallel for a range of inputs and sequentially in a for-loop over the same inputs. Then the decorator version would be something like `@parallel_parametrize`. Does this make sense?

File: scipy/_lib/_testutils.py (hunk 2 above)
Comment (andfoy) [reply scipy/_lib/_testutils.py:396]: Both should be acceptable workloads, for instance, consider the test defined in `test_rgi.py` 

https://github.com/andfoy/scipy/blob/abcf2e4981070a70e9d32ad0ad927ca75ff08a06/scipy/interpolate/tests/test_rgi.py#L703

This test checks for a not mutating attribute of a class that gets called from multiple threads with different arguments, in this case, checking for equality does not make sense

**Comment by rgommers** [comment]
This doesn't quite do what I expected I think - more on that further down. For what this PR currently does, I'm looking at the old and new way of writing a test case:

```python
    def test_concurrency(self):
        P = KroghInterpolator(self.xs, self.ys)

        def worker_fn(_, interp):
            interp(self.xs)

        _run_concurrent_barrier(10, worker_fn, P)
```

now becomes:
```python
    @pytest.fixture
    def default_krogh(self):
        return KroghInterpolator(self.xs, self.ys)

    @run_in_parallel
    def test_concurrency(self, default_krogh):
        return default_krogh(self.xs)
```

or for a more complicated example:

<details>

```python
    def test_concurrency(self):
        points, values = self._get_sample_4d()
        sample = np.array([[0.1 , 0.1 , 1.  , 0.9 ],
                           [0.2 , 0.1 , 0.45, 0.8 ],
                           [0.5 , 0.5 , 0.5 , 0.5 ],
                           [0.3 , 0.1 , 0.2 , 0.4 ]])
        interp = RegularGridInterpolator(points, values, method="slinear")

        # A call to RGI with a method different from the one specified on the
        # constructor, should not mutate it.
        methods = ['slinear', 'nearest']
        def worker_fn(tid, interp):
            spline = interp._spline
            method = methods[tid % 2]
            interp(sample, method=method)
            assert interp._spline is spline

        _run_concurrent_barrier(10, worker_fn, interp)
```

now becomes:
```python
    @pytest.fixture
    def rgi_4d(self):
        points, values = self._get_sample_4d()
        sample = np.array([[0.1 , 0.1 , 1.  , 0.9 ],
                           [0.2 , 0.1 , 0.45, 0.8 ],
                           [0.5 , 0.5 , 0.5 , 0.5 ],
                           [0.3 , 0.1 , 0.2 , 0.4 ]])
        interp = RegularGridInterpolator(points, values, method="slinear")
        return interp, sample

    @run_in_parallel(pass_thread_id=True, assert_all_equal=False)
    def test_concurrency(self, rgi_4d, thread_id=0):
        # A call to RGI with a method different from the one specified on the
        # constructor, should not mutate it.
        interp, sample = rgi_4d
        spline = interp._spline

        methods = ['slinear', 'nearest']
        method = methods[thread_id % 2]
        interp(sample, method=method)
        assert interp._spline is spline
```

</details>

The new version is maybe slightly nicer, but not my much - opinions may vary I suspect. 

* * *

What I actually wanted a decorator for is to make a serial test like `test_somefunc` automatically run in parallel with:
```python
@run_in_parallel
@def_somefunc():
    ...
```
The main goal being that we can smoke out whole classes of errors _without_ having to write new custom `test_concurrency` functions for every bit of functionality. Which is why I pointed out that scikit-image does this already (see [example](https://github.com/scikit-image/scikit-image/blob/0e97db25f3638b923be9b470fc5232403b4ffa95/skimage/transform/tests/test_warps.py#L72-L73) and [the decorator](https://github.com/scikit-image/scikit-image/blob/ff3dd9468bd10461363e1f7fdba9b12409fba216/skimage/_shared/testing.py#L206C1-L206C60)) 

What this PR does is allow for passing the same input objects to a function. What I had in mind was calling the same functionality in parallel. The former may catch some extra issues, but the latter will catch most thread-safety issues that we worry about as well. 

I suggest assessing whether this PR is still useful (curious if other maintainers prefer the old or new test style here). And prioritizing a separate decorator which is more like an enhanced version of the scikit-image one.

**Comment by andfoy** [comment]
This decorator can be invoked to launch the same test in parallel, when used without arguments. The porting done here accounts for previous instances where results where compared.

However, when used as `run_in_parallel`, without parameters, it will run the test in parallel.

**Comment by rgommers** [comment]
Thanks @andfoy, good to hear. That wasn't/isn't obvious from the diff. 

> Maybe an example is missing in the docstring to account for the base case.

It'd be great to make this the primary use case in the docstring then, and add both an example and a few usages in the test suite.

**Comment by andfoy** [comment]
@ev-br @rgommers [a3f42b3](https://github.com/scipy/scipy/pull/21350/commits/a3f42b321f8af612f02467b5de3beaf02f20ee5d) addresses several of the review comments: 

* `use_barrier` was removed as it does not make sense to try to run tests in parallel without any concurrency guarantee
* `assert_all_equal` was renamed to `assert_all_close`, denoting that `assert_allclose` is used underneath
* `**allclose_kwargs` was added to pass keyword arguments to  `assert_allclose`
* `assert_all_close` kwarg was set to  `False`by default, this is done to make sure that the base case required on https://github.com/scipy/scipy/pull/21350#issuecomment-2286054256 takes place by default

**Comment by rgommers** [comment]
Thanks @andfoy. I did a bit of testing with a plain `@run_in_parallel`. It seems to work well so far. One thing that would be good to improve is the tracebacks on failure. If you just take a simple serial test and add an `AssertionError` or some other error, you get a simple traceback like:
```
_________________________________________________________________________________________ TestLinkage.test_linkage_tdist[numpy] _________________________________________________________________________________________
scipy/cluster/tests/test_hierarchy.py:96: in test_linkage_tdist
    raise AssertionError('some error')
E   AssertionError: some error
        method     = 'single'
        self       = <scipy.cluster.tests.test_hierarchy.TestLinkage object at 0x11ccba2d0>
        xp         = <module 'numpy' from '/Users/rgommers/code/pixi-dev/scipy/.pixi/envs/default/lib/python3.12/site-packages/numpy/__init__.py'>
================================================================================================ short test summary info ================================================================================================
FAILED scipy/cluster/tests/test_hierarchy.py::TestLinkage::test_linkage_tdist[numpy] - AssertionError: some error
``` 

While with the decorator you get to see the `pytest` internals in the traceback:

<details>

```
_________________________________________________________________________________________ TestLinkage.test_linkage_tdist[numpy] _________________________________________________________________________________________
../../../../../.pixi/envs/default/lib/python3.12/site-packages/_pytest/runner.py:341: in from_call
    result: TResult | None = func()
        cls        = <class '_pytest.runner.CallInfo'>
        duration   = 0.0035427911207079887
        excinfo    = <ExceptionInfo PytestUnhandledThreadExceptionWarning('Exception in thread Thread-21 (closure)\n\nTraceback (most recen....py", line 97, in test_linkage_tdist\n    raise AssertionError(\'some error\')\nAssertionError: some error\n') tblen=9>
        func       = <function call_and_report.<locals>.<lambda> at 0x11bfd59e0>
        precise_start = 3465200.59691175
        precise_stop = 3465200.600454541
        reraise    = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)
        result     = None
        start      = 1723649559.858759
        stop       = 1723649559.862302
        when       = 'call'
../../../../../.pixi/envs/default/lib/python3.12/site-packages/_pytest/runner.py:242: in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
        item       = <Function test_linkage_tdist[numpy]>
        kwds       = {}
        runtest_hook = <HookCaller 'pytest_runtest_call'>
../../../../../.pixi/envs/default/lib/python3.12/site-packages/pluggy/_hooks.py:513: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
        firstresult = False
        kwargs     = {'item': <Function test_linkage_tdist[numpy]>}
        self       = <HookCaller 'pytest_runtest_call'>
../../../../../.pixi/envs/default/lib/python3.12/site-packages/pluggy/_manager.py:120: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
        firstresult = False
        hook_name  = 'pytest_runtest_call'
        kwargs     = {'item': <Function test_linkage_tdist[numpy]>}
        methods    = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/Users/rgommers/code/pixi-dev/scipy/.pixi/envs/...ure=None>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x119a127e0>>, ...]
        self       = <_pytest.config.PytestPluginManager object at 0x106ba0110>
../../../../../.pixi/envs/default/lib/python3.12/site-packages/_pytest/threadexception.py:92: in pytest_runtest_call
    yield from thread_exception_runtest_hook()
../../../../../.pixi/envs/default/lib/python3.12/site-packages/_pytest/threadexception.py:82: in thread_exception_runtest_hook
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
E   pytest.PytestUnhandledThreadExceptionWarning: Exception in thread Thread-21 (closure)
E
E   Traceback (most recent call last):
E     File "/Users/rgommers/code/pixi-dev/scipy/.pixi/envs/default/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
E       self.run()
E     File "/Users/rgommers/code/pixi-dev/scipy/.pixi/envs/default/lib/python3.12/threading.py", line 1012, in run
E       self._target(*self._args, **self._kwargs)
E     File "/Users/rgommers/code/pixi-dev/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/_testutils.py", line 427, in closure
E       results.append(fn(*args, **kwargs))
E                      ^^^^^^^^^^^^^^^^^^^
E     File "/Users/rgommers/code/pixi-dev/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/cluster/tests/test_hierarchy.py", line 97, in test_linkage_tdist
E       raise AssertionError('some error')
E   AssertionError: some error
        cm         = <_pytest.threadexception.catch_threading_exception object at 0x11c725f40>
        msg        = 'Exception in thread Thread-21 (closure)\n\nTraceback (most recent call last):\n  File "/Users/rgommers/code/pixi-dev/..._hierarchy.py", line 97, in test_linkage_tdist\n    raise AssertionError(\'some error\')\nAssertionError: some error\n'
        thread_name = 'Thread-21 (closure)'
```

</details>

**Comment by andfoy** [comment]
After 46cd97c, the traceback looks like the following:

```
scipy/interpolate/tests/test_bsplines.py::TestNdBSpline::test_concurrency FAILED

====================================================================================== FAILURES ======================================================================================
___________________________________________________________________________ TestNdBSpline.test_concurrency ___________________________________________________________________________

self = <scipy.interpolate.tests.test_bsplines.TestNdBSpline object at 0x4c39e7ea5e0>, random_3d_ndbspline = <scipy.interpolate._ndbspline.NdBSpline object at 0x4c3a1194510>

    @run_in_parallel(assert_all_close=True)
    def test_concurrency(self, random_3d_ndbspline):
        xi = np.c_[[1, 1.5, 2],
                   [1.1, 1.6, 2.1],
                   [0.9, 1.4, 1.9]]
>       raise ValueError('This is an exception')
E       ValueError: This is an exception

random_3d_ndbspline = <scipy.interpolate._ndbspline.NdBSpline object at 0x4c3a1194510>
self       = <scipy.interpolate.tests.test_bsplines.TestNdBSpline object at 0x4c39e7ea5e0>
xi         = array([[1. , 1.1, 0.9],
       [1.5, 1.6, 1.4],
       [2. , 2.1, 1.9]])

scipy/interpolate/tests/test_bsplines.py:2446: ValueError
============================================================================== short test summary info ===============================================================================
FAILED scipy/interpolate/tests/test_bsplines.py::TestNdBSpline::test_concurrency - ValueError: This is an exception
```

Right now this only displays the first exception that comes out from a thread, if multiple exceptions were to appear on different threads, then each would appear once the previous one is fixed. I tried using `ExceptionGroup`, however, the traceback ends up still being verbose and including pytest source files:

```
scipy/interpolate/tests/test_bsplines.py::TestNdBSpline::test_concurrency FAILED

====================================================================================== FAILURES ======================================================================================
___________________________________________________________________________ TestNdBSpline.test_concurrency ___________________________________________________________________________
  + Exception Group Traceback (most recent call last):
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/runner.py", line 341, in from_call
  |     result: TResult | None = func()
  |                              ~~~~^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/runner.py", line 242, in <lambda>
  |     lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
  |             ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_hooks.py", line 513, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |            ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_manager.py", line 120, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |            ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 182, in _multicall
  |     return outcome.get_result()
  |            ~~~~~~~~~~~~~~~~~~^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_result.py", line 100, in get_result
  |     raise exc.with_traceback(exc.__traceback__)
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/threadexception.py", line 92, in pytest_runtest_call
  |     yield from thread_exception_runtest_hook()
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
  |     yield
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/unraisableexception.py", line 95, in pytest_runtest_call
  |     yield from unraisable_exception_runtest_hook()
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
  |     yield
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/logging.py", line 848, in pytest_runtest_call
  |     yield from self._runtest_for(item, "call")
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/logging.py", line 831, in _runtest_for
  |     yield
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/capture.py", line 879, in pytest_runtest_call
  |     return (yield)
  |             ^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/skipping.py", line 257, in pytest_runtest_call
  |     return (yield)
  |             ^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 103, in _multicall
  |     res = hook_impl.function(*args)
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
  |     item.runtest()
  |     ~~~~~~~~~~~~^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/python.py", line 1627, in runtest
  |     self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_hooks.py", line 513, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |            ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_manager.py", line 120, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |            ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 139, in _multicall
  |     raise exception.with_traceback(exception.__traceback__)
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 122, in _multicall
  |     teardown.throw(exception)  # type: ignore[union-attr]
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^
  |   File "/home/andfoy/Documentos/Quansight/nogil/pytest-run-parallel/src/pytest_run_parallel/plugin.py", line 62, in pytest_pyfunc_call
  |     return (yield)
  |             ^^^^^
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/pluggy/_callers.py", line 103, in _multicall
  |     res = hook_impl.function(*args)
  |   File "/home/andfoy/.pyenv/versions/3.13.0rc1t/lib/python3.13t/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
  |     result = testfunction(**testargs)
  |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 453, in inner
  |     raise exc
  | ExceptionGroup: one or more threads raised exceptions (10 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 3 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 4 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 5 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 6 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 7 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 8 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 9 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +---------------- 10 ----------------
    | Traceback (most recent call last):
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/_lib/_testutils.py", line 430, in closure
    |     results.append(fn(*args, **kwargs))
    |                    ~~^^^^^^^^^^^^^^^^^
    |   File "/home/andfoy/Documentos/Quansight/nogil/scipy/scipy/interpolate/tests/test_bsplines.py", line 2446, in test_concurrency
    |     raise ValueError('This is an exception')
    | ValueError: This is an exception
    +------------------------------------
============================================================================== short test summary info ===============================================================================
FAILED scipy/interpolate/tests/test_bsplines.py::TestNdBSpline::test_concurrency - ExceptionGroup: one or more threads raised exceptions (10 sub-exceptions)

```

**Comment by ev-br** [comment]
Where does this PR stand w.r.t. https://github.com/Quansight-Labs/pytest-run-parallel/ --- are these independent / orthogonal, or does the plugin supersede this PR, or...?

**Comment by ev-br** [comment]
So the plugin in the way forward, and this PR will be... eventually redone using the plugin?

**Comment by andfoy** [comment]
I'm not sure, according to @rgommers' plan (detailed at https://github.com/Quansight-Labs/free-threaded-compatibility/issues/56), the decorator and the plugin are separate items that should complement each other (as opposed for one to be the replacement of the other)

**Comment by ev-br** [comment]
OK, let me rephrase: how does this compare to https://github.com/scipy/scipy/pull/21496? Both seem to bring threading-related testing primitives.

**Comment by andfoy** [comment]
The difference here is that this decorator helps with explicit parallel testing as opposed of taking an already existing test and running in parallel, which are two different approaches

**Comment by ev-br** [comment]
Exactly. 
So my confusion is why we need two different approaches in scipy. Would one or the other be sufficient?

**Comment by andfoy** [comment]
The way to think about it is that the decorator helps to test concurrent execution  for singleton instances across tests as opposed to n threads creating different instances of the same class.

The decorator will cover those cases where the plugin is not enough. Mostly singleton object instances at this time.

**Comment by ev-br** [comment]
Thanks. 
So, you're saying that in the long run we'd want to have both ways of testing then?
In this case it'd be great to see the rollout plan.

**Comment by rgommers** [comment]
@andfoy has this PR been fully superseded by `pytest-run-parallel`, or is there still anything we want from this you think?

**Comment by andfoy** [comment]
We can close this one, as the future CI job will run the test suite in parallel
